client/network/src/protocol.rs

```diff
@@ -1659,6 +1694,14 @@ impl<B: BlockT, H: ExHashT> NetworkBehaviour for Protocol<B, H> {
 		};
 
 		if let CustomMessageOutcome::None = outcome {
+			// Check if there is any block announcement validation finished.
+			while let Poll::Ready(result) = self.sync.poll_block_announce_validation(cx) {
```

Comment:
Why is this nested within `if let CustomMessageOutcome`? Why not move this towards the beginning of the method doing `self.pending_messages.push_back(NetworkBehaviourAction::GenerateEvent(outcome))` instead of returning.

Reply:
Because you could have added a block announce above, which in most of the times will directly be finished when `poll` being called here.

Reply:
> you could have added a block announce above

That should lead to a wake-up of the underlying task, which then results in this poll function being called again which then calls `self.sync.poll_block_announce_validation`. Depending on the ordering of these calls seems error prone to me.

Reply:
To wake up the underlying task, I need to poll a future at least once. 

And honestly, the same structure already exists above with these `if let Some() { return Poll::Ready() }`.
I'm also not 100% happy with the solution, but I don't see the advantage of moving this above. What would change?